src/core/pratter.ml

```diff
@@ -0,0 +1,131 @@
+(* FIXME: use opam version instead *)
+(** This modules defines a functor whose image is a parser for terms with
+    applications, binary and unary operators. These terms are specified in the
+    argument of the functor.
+
+    The algorithm implemented is an extension of the Pratt parser. The Sunting
+    Yard algorithm could also be used.
+    @see <https://dev.to/jrop/pratt-parsing>
+    @see <https://effbot.org/zone/simple-top-down-parsing.htm> *)
+
+(** Associativity of an operator. *)
+type associativity =
+  | Left
+      (** If [+] is a left associative operator, [x + y + z] is parsed [(x +
+          y) + z]. *)
+  | Right
+      (** If [+] is a right associative operator, [x + y + z] is parsed [x +
+          (y + z)]. *)
+
+type priority = float
+(** Priority of operators. If [*] has a higher priority than [+], than [x + y
+    * z] is parsed [x + (y * z)]. *)
+
+(** Types and utilities on terms that are to be Pratt parsed. *)
+module type SUPPORT = sig
+  type term
+  (** The main type of terms, that contains symbols, applications, binary and
+      unary operators. *)
+
+  type table
+  (** The table is used to store available operators. *)
+
+  val get_unary : table -> term -> priority option
+  (** [get_unary tbl t] returns the priority, or binding power of operator
+      identified [t] if it is a unary operator, or [None]. *)
+
+  val get_binary : table -> term -> (priority * associativity) option
+  (** [get_binary tbl t] returns the priority (or binding power) and
+      associativity of operator [t], or [None] if [t] isn't a binary
+      operator. *)
```

Comment:
Why not having a single function
```
type op = Una | Bin of associativity
val get_sym : table -> term -> (op * priority) option
```
?

Reply:
Remark: it may look like this forbids to have identifiers that can be both unary and binary symbols depending on the number of arguments they are applied to, like - for instance. But we could later implement this as well by modifying the sign type and the Sign.find function.

Reply:
It may require quite a change in the pratt parser...

Reply:
Are you talking of my first question or my second remark?

Reply:
The second indeed.